### DISCUSSION TITLE
Circuit problem "django.db.utils.IntegrityError"

### DISCUSSION THREAD

**Discussion by dejantep** [netbox-community/netbox#8166] [Q&A]
I have a circuit connected to interface and visible in interface overview of a device.
But when go to circuit detail page there are no terminations visible. If i try to add termination A i get following error

duplicate key value violates unique constraint "circuits_circuittermination_circuit_id_term_side_b13efd0e_uniq"
DETAIL:  Key (circuit_id, term_side)=(81, A) already exists.

Python version: 3.6.9
NetBox version: 2.11.12

How would one correct this?

**Comment by candlerb** [comment]
I don't understand.  How did you get a "circuit connected to interface" without using a circuit termination?

Otherwise, I wonder if it's a permissions thing.  If you login to Netbox using a superuser account, are the terminations visible?

**Comment by dejantep** [comment]
I'm a superuser and circuit was connected a while ago. Log retention is 90 days so there is nothing to look for in change log
I'm not sure what could have happened.

Reason i stumbled upon this is that i updated circuit type on this particular circuit.

If i just disconnect from the interface will that remove values that django is complaining about?

**Reply by candlerb** [reply]
I think you should first try to understand what's going on.

Can you give a screenshot of the circuit? (`/circuits/circuits/NNN/`)

Does it show "Termination - A Side: None" and "Termination Z Side: None"?  Or something else?

**Reply by dejantep** [reply]
it shows "Termination - A Side: None" and "Termination Z Side: None"

**Reply by candlerb** [reply]
OK.  Now try a direct database query: under Ubuntu I get a postgres SQL shell using

```
sudo -u postgres psql netbox
```

Then show the results of these SQL queries:

```
select id,termination_a_id,termination_z_id from circuits_circuit;
select id,circuit_id,term_side,cable_id from circuits_circuittermination;
```

(Note: I'm checking these with Netbox 3.1, but I don't think Netbox 2.11 was significantly different here)

There seems to be some duplication in the data model here, with the termination IDs stored within the circuit model, and I wonder if this is leading to the inconsistency.

**Reply by dejantep** [reply]
thankj you for thiis. Here's relevant row from first query

 id  | termination_a_id | termination_z_id  
81 |                  |                 

And here it is from the second query

 id  | circuit_id | term_side | cable_id 
 106 |         81 | A         |      427

**Reply by dejantep** [reply]
i can now say that there is no cable with id 427

**Reply by candlerb** [reply]
I guess the problem is that termination_a_id is null in the circuit, even though the circuit does have a termination_a.  Try this:

```
update circuits_circuit set termination_a_id=106 where id=81;
```

**Reply by candlerb** [reply]
Also try these queries to see if there are any more such anomalies:

```
select c.id,c.termination_a_id,ta.id from circuits_circuit c
  left join circuits_circuittermination ta on ta.circuit_id=c.id and ta.term_side='A'
  where c.termination_a_id is distinct from ta.id;

select c.id,c.termination_z_id,tz.id from circuits_circuit c
  left join circuits_circuittermination tz on tz.circuit_id=c.id and tz.term_side='Z'
  where c.termination_z_id is distinct from tz.id;
```

**Reply by candlerb** [reply]
> i can now say that there is no cable with id 427

Hmm, that means there's another problem: there's a missing foreign key constraint which would have prevented this.

In the PostgreSQL shell again, try this:

```
\d circuits_circuittermination
```

In one Netbox system (which was built years ago and migrated to v3.1.2) I get:

```
...
Foreign-key constraints:
    "circuits_circuitterm__link_peer_type_id_4c96d182_fk_django_co" FOREIGN KEY (_link_peer_type_id) REFERENCES django_content_type(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuitterm_provider_network_id_b0c660f1_fk_circuits_" FOREIGN KEY (provider_network_id) REFERENCES circuits_providernetwork(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_circuit_id_257e87e7_fk" FOREIGN KEY (circuit_id) REFERENCES circuits_circuit(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_site_id_e6fe5652_fk_dcim_site_id" FOREIGN KEY (site_id) REFERENCES dcim_site(id) DEFERRABLE INITIALLY DEFERRED
```

In another, which was a test system built more recently as v3.0 and then migrated, I get:

```
...
Foreign-key constraints:
    "circuits_circuitterm__link_peer_type_id_4c96d182_fk_django_co" FOREIGN KEY (_link_peer_type_id) REFERENCES django_content_type(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuitterm_circuit_id_257e87e7_fk_circuits_" FOREIGN KEY (circuit_id) REFERENCES circuits_circuit(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuitterm_provider_network_id_b0c660f1_fk_circuits_" FOREIGN KEY (provider_network_id) REFERENCES circuits_providernetwork(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_cable_id_35e9f703_fk_dcim_cable_id" FOREIGN KEY (cable_id) REFERENCES dcim_cable(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_site_id_e6fe5652_fk_dcim_site_id" FOREIGN KEY (site_id) REFERENCES dcim_site(id) DEFERRABLE INITIALLY DEFERRED
```

So my main system is missing the `circuits_circuittermination_cable_id` constraint, and I guess yours is too.  I think we are both suffering from "migrations drift".

**Reply by candlerb** [reply]
BTW there's a possible way to test for this, using [sqldiff](https://django-extensions.readthedocs.io/en/latest/sqldiff.html), but it doesn't appear to work very well.

```
cd /opt/netbox
. venv/bin/activate
pip install django-extensions
# now add
#        'django_extensions',
# to the end of `INSTALLED_APPS` in `/opt/netbox/netbox/netbox/settings.py`
python netbox/manage.py sqldiff circuits
python netbox/manage.py sqldiff dcim
```

This is a bit of a frig because you're not supposed to touch `settings.py`, so put it back after you're done.

My results:

```
(venv) root@netbox3:/opt/netbox# python netbox/manage.py sqldiff circuits
-- No differences
(venv) root@netbox3:/opt/netbox# python netbox/manage.py sqldiff dcim
BEGIN;
-- Application: dcim
-- Model: Interface
-- Comment: Unknown database type for field 'mac_address' (829)
ALTER TABLE "dcim_interface"
	ALTER "mac_address" TYPE macaddr;
-- Comment: Unknown database type for field 'wwn' (774)
ALTER TABLE "dcim_interface"
	ALTER "wwn" TYPE macaddr8;
-- Model: InventoryItem
CREATE INDEX "dcim_inventoryitem_tree_id_lft_8a1329b5_idx"
	ON "dcim_inventoryitem" ("tree_id", "lft");
-- Model: Region
ALTER TABLE "dcim_region"
	DROP CONSTRAINT "dcim_region_parent_name";
ALTER TABLE "dcim_region"
	DROP CONSTRAINT "dcim_region_parent_slug";
DROP INDEX "dcim_region_name";
DROP INDEX "dcim_region_slug";
CREATE INDEX "dcim_region_tree_id_lft_cc93ebcb_idx"
	ON "dcim_region" ("tree_id", "lft");
-- Model: SiteGroup
ALTER TABLE "dcim_sitegroup"
	DROP CONSTRAINT "dcim_sitegroup_parent_name";
ALTER TABLE "dcim_sitegroup"
	DROP CONSTRAINT "dcim_sitegroup_parent_slug";
DROP INDEX "dcim_sitegroup_name";
DROP INDEX "dcim_sitegroup_slug";
CREATE INDEX "dcim_sitegroup_tree_id_lft_2a27eab4_idx"
	ON "dcim_sitegroup" ("tree_id", "lft");
-- Model: Location
ALTER TABLE "dcim_location"
	DROP CONSTRAINT "dcim_location_parent_name";
ALTER TABLE "dcim_location"
	DROP CONSTRAINT "dcim_location_parent_slug";
DROP INDEX "dcim_location_name";
DROP INDEX "dcim_location_slug";
CREATE INDEX "dcim_location_tree_id_lft_5b61f5ca_idx"
	ON "dcim_location" ("tree_id", "lft");
COMMIT;
```

So sadly, it didn't pick up the missing index on `circuits_circuittermination`.  Furthermore, it claims to find some issues with dcim, but nothing relevant to this.  I don't think I trust it anyway, because even the newer system also reports the same problems:

```
(venv) root@netbox-test:/opt/netbox# python netbox/manage.py sqldiff circuits
-- No differences
(venv) root@netbox-test:/opt/netbox# python netbox/manage.py sqldiff dcim
BEGIN;
-- Application: dcim
-- Model: Interface
-- Comment: Unknown database type for field 'mac_address' (829)
ALTER TABLE "dcim_interface"
	ALTER "mac_address" TYPE macaddr;
-- Comment: Unknown database type for field 'wwn' (774)
ALTER TABLE "dcim_interface"
	ALTER "wwn" TYPE macaddr8;
-- Model: InventoryItem
CREATE INDEX "dcim_inventoryitem_tree_id_lft_8a1329b5_idx"
	ON "dcim_inventoryitem" ("tree_id", "lft");
-- Model: Region
ALTER TABLE "dcim_region"
	DROP CONSTRAINT "dcim_region_parent_name";
ALTER TABLE "dcim_region"
	DROP CONSTRAINT "dcim_region_parent_slug";
DROP INDEX "dcim_region_name";
DROP INDEX "dcim_region_slug";
CREATE INDEX "dcim_region_tree_id_lft_cc93ebcb_idx"
	ON "dcim_region" ("tree_id", "lft");
-- Model: SiteGroup
ALTER TABLE "dcim_sitegroup"
	DROP CONSTRAINT "dcim_sitegroup_parent_name";
ALTER TABLE "dcim_sitegroup"
	DROP CONSTRAINT "dcim_sitegroup_parent_slug";
DROP INDEX "dcim_sitegroup_name";
DROP INDEX "dcim_sitegroup_slug";
CREATE INDEX "dcim_sitegroup_tree_id_lft_2a27eab4_idx"
	ON "dcim_sitegroup" ("tree_id", "lft");
-- Model: Location
ALTER TABLE "dcim_location"
	DROP CONSTRAINT "dcim_location_parent_name";
ALTER TABLE "dcim_location"
	DROP CONSTRAINT "dcim_location_parent_slug";
DROP INDEX "dcim_location_name";
DROP INDEX "dcim_location_slug";
CREATE INDEX "dcim_location_tree_id_lft_5b61f5ca_idx"
	ON "dcim_location" ("tree_id", "lft");
COMMIT;
```

**Reply by DanSheps** [reply]
I think the reasons for issues with DCIM are that we index them because we validate their uniqueness but do that in code.

**Reply by candlerb** [reply]
sqldiff seems to think that the definitions seen from talking to the database itself do not match the definitions in the models.  But it could be a bug in sqldiff.

**Reply by dejantep** [reply]
Hello and thank you once again for your valuable inputs.

I did check for other anomalies and i'v got 3 rows on termination A. Nothing on termination Z
 id | termination_a_id | id  
----+------------------+-----
 81 |                  | 106
 82 |                  | 107
 87 |                  | 117 

and regarding foreign-key constraints i have this. My system has been around since version 2.7.x 

```
    "circuits_circuitterm__cable_peer_type_id_bd122156_fk_django_co" FOREIGN KEY (_cable_peer_type_id) REFERENCES django_content_type(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuitterm_provider_network_id_b0c660f1_fk_circuits_" FOREIGN KEY (provider_network_id) REFERENCES circuits_providernetwork(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_circuit_id_257e87e7_fk" FOREIGN KEY (circuit_id) REFERENCES circuits_circuit(id) DEFERRABLE INITIALLY DEFERRED
    "circuits_circuittermination_site_id_e6fe5652_fk_dcim_site_id" FOREIGN KEY (site_id) REFERENCES dcim_site(id) DEFERRABLE INITIALLY DEFERRED
```

I cannot do that last part with sqldiff but thank you anyway.

I would like to fix those 3 errored rows. Would your update query solve that?

**Reply by candlerb** [reply]
Yes, you should be able to modify that update as appropriate to fill in the missing termination_a_id on each circuit.

However if the circuit_id points to a non-existent cable, that won't be fixed.  To check for those:

```
select ct.id,ct.cable_id from circuits_circuittermination ct
  left join dcim_cable dc on ct.cable_id = dc.id
  where ct.cable_id is not null and dc.id is null;
```

I'm not sure what would be the best fix for those, but I think just setting the cable_id to null would be reasonable (on the assumption that the cable has been deleted).

Once that's done, in principle you can add the foreign key constraint by hand.  It would however be nice to know how the database got into that bad state.

**Reply by dejantep** [reply]
Thanks a lot. i managed to correct it.

Happy holidays!!
